### ISSUE TITLE
Ability to omit documentation from rendered source

### ISSUE THREAD

**Issue by beanaroo** [mkdocstrings/mkdocstrings#249]
**Is your feature request related to a problem? Please describe.**
`show_source` is great, but a lot of content on the page is duplicated. This creates quite a bit of noise where a lot of code is documented.

**Describe the solution you'd like**
Being able to omit documentation from rendered source code for functions/classes/methods.

**Additional context**
![image](https://user-images.githubusercontent.com/1748909/109408293-f2f02800-79ec-11eb-9b3f-25c7c2bc15b8.png)

**Comment by oprypin** [comment]
I have a good workaround with CSS that works now. Just hide those elements as per Pygments highlighting.

You may notice side effects, one I already see is that it also removes blank lines...
It's because I want to hide blank lines after I hid the doc element, but there's no way to discern which lines those are.

```css
.doc-contents details .highlight code {
  line-height: 0;
}
.doc-contents details .highlight code>* {
  line-height: initial;
}
.doc-contents details .highlight code>.sd {  /* Literal.String.Doc */
  display: none;
}
```
You can add it with MkDocs `extra_css`

**Comment by pawamoy** [comment]
That's a thought I had as well @beanaroo, thanks for bringing it here :slightly_smiling_face: 
I have to admit that sometimes it's nice to see how the docstring was written, but yeah, most of the time it just takes a lot of space for nothing.

Thanks for the trick @oprypin! This should definitely be included in a wiki or in the docs :smile:

**Comment by thatlittleboy** [comment]
> The [new handler](https://mkdocstrings.github.io/handlers/overview/#about-the-python-handlers) knows the starting and ending locations of both the object and its docstring, so I believe hiding the docstring in the source code block will be possible without CSS tricks 😄

@pawamoy  I've tried to make some headway towards this, here's my idea (it kinda involves making changes to multiple repos, which is why I'm not presenting this in a PR):

### My idea

As you prompted, griffe knows about docstring start/end, so we can remove docstring from source before rendering in the template.

In [`griffe/dataclasses.py`](https://github.com/mkdocstrings/griffe/blob/b4f54e49ee3c52d33ae54bf3ef87b31ae0b40ff9/src/griffe/dataclasses.py#L663):
```python
class Object(...):

    # ... [truncated]

    @cached_property
    def source_no_docstring(self) -> str:
        """Return the source code of this object with the docstring stripped out.

        Returns:
            The source code with no docstring.
        """
        if not self.has_docstring:
            return self.source
        docstring = self.docstring
        lines = self.lines
        return dedent("\n".join(
            lines[:docstring.lineno - self.lineno] + lines[docstring.endlineno - self.lineno + 1:]
        ))
```
And in [`mkdocstrings-python:mkdocstrings_handlers/python/templates/material/_base/function.html`](https://github.com/mkdocstrings/python/blob/4ce0fc824fc36cadcf2b70cc72b627bdaa5abd65/src/mkdocstrings_handlers/python/templates/material/_base/function.html#L68)
```jinja
    {% if config.show_source and function.source %}
      <details class="quote">
        <summary>Source code in <code>{{ function.relative_filepath }}</code></summary>
        {# Add a new global configuration option to control whether to strip docstring from source or not #}
        {% if config.strip_docstring_from_source %}
          {{ function.source_no_docstring|highlight(language="python", linestart=function.lineno, linenums=True) }}
        {% else %}
          {{ function.source|highlight(language="python", linestart=function.lineno, linenums=True) }}
        {% endif %}
      </details>
```

With `strip_docstring_from_source: no` set in `mkdocs.yml` (i.e., effectively the same as current status quo)

<img width="400" src="https://user-images.githubusercontent.com/30731072/213871431-43bc34fb-6332-4d89-8841-d453c192ca59.png">

With `strip_docstring_from_source: yes` set in `mkdocs.yml`

<img width="400" src="https://user-images.githubusercontent.com/30731072/213871500-d7688120-aa6d-4c84-9d95-4caf90eb88a6.png">

### Problem

**Qn:**: There's just one problem, the line numbers are wrong after stripping docstrings, and I can't quite figure out a way to get around this -- it probably needs upstream support, right? From pymdownx or pygments itself?

### Other comments

The above is just a **very rough** Proof-of-concept, maybe there's a cleaner way to organize / implement the code, but the main issue I see is still with the line numbering.

Of course, if this were to be implemented, there's alot more to do:
- `class.html` also has `show_source`, so that file needs a similar treatment to `function.html`.
- not sure if other themes other than `Material` need support (haven't looked into it, maybe you can enlighten me)
- documentation for the new configuration option

etc..

@pawamoy What do you think, is this even possible? I'll really love to have this supported natively rather than hacking the CSS. 😅

**Comment by pawamoy** [comment]
@thatlittleboy that would be the way to implement that, yes :smile: 
About the then wrong line numbers, I think it needs support from upstream indeed. I don't think it's currently possible to specify multiple ranges of numbers for groups of lines, or something like that (@facelessuser?). Ideally we would be able to specify line numbers like this:

````md
```python linenums="75-79 92"
# this line would be numbered 75
# this one 76, etc., up to 79
# then it would jump to 92 and continue
```
````

**Comment by thatlittleboy** [comment]
Great, thanks for the confirmation~ Lemme try to understand what needs to be patched upstream (pymdownx) and will create an issue there if necessary.

**Comment by facelessuser** [comment]
Yes, pymdownx mainly lets Pygments decide this. At this time, I have no plans to add additional line logic on top of what Pygments already does. It would be too fragile if Pygments ever changed anything.

**Comment by thatlittleboy** [comment]
@facelessuser yup, fair enough, thanks for chipping in! :) I take it that if `pygments` offers this functionality, `pymdownx` will too?

**Comment by facelessuser** [comment]
If they were to expose it, I'd have to look at **how** it is exposed. I'd imagine there would have to be a syntax update to allow people to specify non-contiguous chunks of lines.

The only thing you could do now is to create a special custom fence that sends content through the highlighter, and then after the fact, removes the lines you don't want to see. I guess you could inject a line with ellipses to show a removed chunk of lines if you want 🤷🏻 .

**Comment by pawamoy** [comment]
@thatlittleboy I had another idea. In mkdocstrings, we provide a `highlight` filter which calls the pymdownx highlighter directly. We could add options in this filter to remove ranges of lines. Not sure how hard it would be.

**Comment by patrick-egenlauf** [comment]
@pawamoy I have tried to use the suggested workaround of commit [`c7b27fb`](https://github.com/mkdocstrings/mkdocstrings/commit/c7b27fbefaf0e98b4b9d20bfa3ab86c35d756fdd) but observe the behavior shown in the example below.

![example](https://github.com/user-attachments/assets/a21f7b51-edc3-49da-a351-b9c4dcbb6876)

The docstring is removed, but the line numbers are still displayed. Am I missing something?

**Comment by pawamoy** [comment]
@patrick-egenlauf that might depend on how you configure line numbers to be shown. Could you share your pymdownx.highlight config? 🤔 Or the config or other relevant extensions.

**Comment by patrick-egenlauf** [comment]
@pawamoy  I have the following configuration in my `mkdocs.yml` file:

```yaml
markdown_extensions:
  - pymdownx.tabbed:
      alternate_style: true
  - pymdownx.highlight:
      use_pygments: true
      pygments_lang_class: true
```

Apart from that, I do not configure pymdownx.highlight.

**Comment by pawamoy** [comment]
Thanks! I'm not sure what happens. You could check the generated HTML to see if the CSS selectors are still relevant 🤷 Don't hesitate to share this HTML by the way
